### DISCUSSION TITLE
Renovate JSON schema allows invalid package rules

### DISCUSSION THREAD

**Discussion by 1999** [renovatebot/renovate#26484] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

N/A

### If you're self-hosting Renovate, select which platform you are using.

Bitbucket Cloud (bitbucket.org)

### Was this something which used to work for you, and then stopped?

I have not seen this working

### Describe the problem

We use recommended URL as a $schema: https://docs.renovatebot.com/renovate-schema.json
The problem is anything is treated as a valid package rule.
Am I missing something or is it's designed like this?

![image](https://github.com/renovatebot/renovate/assets/73191/f727e66d-a5f4-4d78-89fb-f77ba271de2a)

### Relevant debug logs

<details><summary>Logs</summary>

```

Copy/paste the relevant log(s) here, between the starting and ending backticks

```

</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by rarkins** [comment, marked as answer]
It's not possible for the JSON schema to validate everything - for this we have our native validation. But if you can figure out how to make it more restrictive for packageRules specifically, you're welcome to submit a PR

**Comment by 1999** [comment]
> It's not possible for the JSON schema to validate everything

Agreed but if you're not relying on JSON schema, maybe it's not worth mentioning it in the docs as a "$schema"?

> you're welcome to submit a PR

Can't promise but I will give it a shot when I have time.

**Reply by 1999** [reply]
PR up: https://github.com/renovatebot/renovate/pull/26560

**Comment by 1999** [comment]
@rarkins a bit of a generic question: do you mean that it is impossible to define the configuration in JSON schema (would be great to understand more and see an example) or you're more concerned about the errors being very cryptic? On a side note, we use Typescript types as a source of truth in one of our project and generate JSON schema from these types of you're interested.

**Reply by rarkins** [reply]
What I mean is that currently there is not enough authoritative source data to generate the JSON schema with such accuracy. The best place it should be added is https://github.com/renovatebot/renovate/blob/main/lib/config/options/index.ts and it would need to be me or a core contributor who does it (too hard for an external contributor). Then this source data could be used to generate the schema accurately.

I created https://github.com/renovatebot/renovate/issues/26571 to move this along
